src/credentials/CHIPOperationalCredentials.cpp

```diff
@@ -0,0 +1,240 @@
+/*
+ *
+ *    Copyright (c) 2021 Project CHIP Authors
+ *    All rights reserved.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+/**
+ *    @file
+ *      This file defines data types and objects for modeling and
+ *      working with CHIP Operational Credentials.
+ *
+ */
+
+#ifndef __STDC_LIMIT_MACROS
+#define __STDC_LIMIT_MACROS
+#endif
+
+#include <credentials/CHIPOperationalCredentials.h>
+#include <support/CHIPMem.h>
+#include <support/ReturnMacros.h>
+
+namespace chip {
+namespace Credentials {
+
+using namespace chip::Crypto;
+
+CHIP_ERROR OperationalCredentialSet::Init(uint8_t maxCertsArraySize)
+{
+    VerifyOrReturnError(maxCertsArraySize > 0, CHIP_ERROR_INVALID_ARGUMENT);
+    mOpCreds = reinterpret_cast<ChipCertificateSet *>(chip::Platform::MemoryAlloc(sizeof(ChipCertificateSet) * maxCertsArraySize));
+    VerifyOrReturnError(mOpCreds != nullptr, CHIP_ERROR_NO_MEMORY);
+
+    mOpCredCount         = 0;
+    mMaxCerts            = maxCertsArraySize;
+    mMemoryAllocInternal = true;
+
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR OperationalCredentialSet::Init(ChipCertificateSet * certSetsArray, uint8_t certSetsArraySize)
+{
+    VerifyOrReturnError(certSetsArray != nullptr, CHIP_ERROR_INVALID_ARGUMENT);
+    VerifyOrReturnError(certSetsArraySize > 0, CHIP_ERROR_INVALID_ARGUMENT);
+
+    mOpCredCount         = certSetsArraySize;
+    mOpCreds             = certSetsArray;
```

Comment:
should we clear/detect if mOpCreds is already non-null? same for above init. This is to not allow Init to be called twice.

It may be easier to follow memory ownership if the 'array with/without ownership' would be a separate object type.

Reply:
Thanks, updated.